lib/payload.json

```diff
@@ -20,6 +20,34 @@
       "description": "Speed (m/s)",
       "minimum": 0
     },
+    "depth": {
+      "type": "number",
+      "description": "Depth from the surface (cm)",
+      "minimum": 0
+    },
```

Comment:
Every addition triggers new ideas.... I guess the first definitions here take time and then things fit in more naturally.

About depth; it's actually a position indicator. How about an object with longitude, latitude, altitude (above sea level) and accuracy? Those fields would come from a GPS receiver. That GPS receiver may not be in the same position as the sensor (if the GPS is above ground and the sensor is a meter under ground), so we may want to indicate a (relative) "level" (+/-) from that position. It's also useful for the wind sensor (which may be higher than the GPS receiver).

If we put a more generic level in the payload root (like now), developers may run into problems with combining measurements that are taken at different levels (i.e. wind sensor and a soil moisture sensor in one device). So would it be wise to assume that `depth` is the soil level (always positive, going from surface to ground) and so it should be in `soil`, to allow for an `elevation` in `wind`? Or would we have a generic `level` field in the root and if there are conflicts, require returning an array with different `level`s? I don't think that having both `elevation` and `depth` in the root is a good idea.

Reply:
On this note; does the codec always know the depth in centimeters? Or do we have an alternative scalar value?

Our friends from @decentlab for instance have this DL-SMTP ([datasheet](https://cdn.decentlab.com/download/datasheets/Decentlab-DL-SMTP-datasheet.pdf)) device (which you may know) and it seems to use relative levels. 

BTW, about the soil moisture value that the sensor transmits:

>The soil moisture values are in the range of –5 to +120. The moisture sensors are calibrated to return
0.000 in air and 100.000 in water. Please note that the returned values ARE NOT soil moisture readings
in percents, but simply a scaled frequency unit (SFU). Please refer to AquaCheck documentation for the
conversion of SFU to volumetric water content.

Is this something that requires another field for soil moisture, or do you think that the conversion from this SFU to volumetric water content (%) can be done in the codec? Or would that need to be calibrated by the application?

Reply:
These are good points, I hadn't thought of them before.

Let's see, we already have an array of `measurements` in the normalized payload, would it be too much to ask for each device's normalization function to separate a payload into multiple `measurements` to avoid conflicts? For example:

This decoded payload:
```json
{
  "soilMoistureAtSurface": 11.2,
  "soilMoistureAtLevel6": 15.5,
  "windSpeed": 5.1
}
```

would become:
```json
[
  {
    "soil": {
      "depth": 0,
      "moisture": 11.2
    }
  },
  {
    "soil": {
      "depth": 60,
      "moisture": 15.5
    }
  },
  {
    "wind": {
      "elevation": 10,
      "speed": 5.1
    }
  }
]
```

That's leaving too much responsibility to the uplink normalizer function, but this is the first thing that came to my mind to solve this and support every use case possible.

Regarding the conversion units specific to the manufacturer, such us levels or scaled units from the calibration process, I think that those should be handled in the normalizer function as well (assuming that the info needed to make the conversion comes in the payload). The normalized format should only expose units that are "universally" relevant in the domains the data may be used, in the example above: SFU should be converted to volumetric water content (%). Otherwise, we will be leaking manufacturer specific details to applications, the same thing that we already have today with standard payloads.

Reply:
> That's leaving too much responsibility to the uplink normalizer function, but this is the first thing that came to my mind to solve this and support every use case possible.

Yes this is also exactly what I think we should do.

---

> Regarding the conversion units specific to the manufacturer, such us levels or scaled units from the calibration process, I think that those should be handled in the normalizer function as well (assuming that the info needed to make the conversion comes in the payload)

I mean the situation where the end device isn't even aware of the depth. Let's say uncle George sticks a soil moisture sensor in his vinyard and only he knows he deep he put it in.

Just for clarity: I do still think we need a field for depth in centimeters. I just think we also need a field for a more generic "level" (number, starting from 0) that the application can map to a depth based on context.

In the long run, what else we should do in the device repository, is add (installation) state. Each device could declare what's relevant to know for an installed device (like depth), and the values would be part of the `input` to the codecs, so that the codecs can use those values as multipliers or references (if needed). In TTS Console, we could render a form to configure that state per device.

---

> The normalized format should only expose units that are "universally" relevant in the domains the data may be used, in the example above: SFU should be converted to volumetric water content (%). Otherwise, we will be leaking manufacturer specific details to applications, the same thing that we already have today with standard payloads.

Agreed

Reply:
> I mean the situation where the end device isn't even aware of the depth. Let's say uncle George sticks a soil moisture sensor in his vinyard and only he knows he deep he put it in.
> 
> Just for clarity: I do still think we need a field for depth in centimeters. I just think we also need a field for a more generic "level" (number, starting from 0) that the application can map to a depth based on context.

I am not sure if I get what you want here. If an end device doesn't support configuring the depth via downlink or other method, how are we supposed to know in the normalization phase what depth/level has to be assigned to that device? As far as I understand, the depth configuration has to either come in the payload if the device supports it or be handled by the integration completely. The normalization function does not have any kind of state, right? It works only with the input that it receives from the device.

Reply:
It would be a logical value. Just like the DL-SMTP (that has six levels); no depth in centimeters but a number assigned to a sensor. You can't (and don't need to) send a downlink message with the depth in centimeters per logical level.

So I'm more looking after:

```json
[
  {
    "soil": {
      "level": 1,
      "moisture": 25
    }
  },
  {
    "soil": {
      "level": 2,
      "moisture": 32
    }
  }
]
```

In the future, if we have installation state:

```json
{
  "depths": [
    {
      "level": 1,
      "depth": 28
    },
    {
      "level": 2,
      "depth": 41
    }
  ]
}
```

And make that state available to the normalizer, then we can map it. For now, we can't but we need some way to convey non-depth-in-cm-levels I guess.

Reply:
I'll check with the Decentlab guys what their view is on this.

Reply:
Got a reply from Decentlab. Bottomline is that soil measurements oftentimes need calibration and these parameters aren't (currently) available in the payload normalizer. For that we need per-device state. Let's not bother with SFU values in the normalized payload but wait until we can provide enough context to the normalizer for it to compute the VWC from SFU (i.e. the linear conversion parameters).

As for depth, forget about the relative depth: let's not do that indeed. The payload normalizer should have enough state to come up with the depth. I'm gonna add version identifiers (https://github.com/TheThingsNetwork/lorawan-stack/blob/feature/normalized-payload-warnings/api/identifiers.proto#L158-L162) to the normalizer's input, so that if there are different models with different depth configurations (like DL-SMTP), the normalizer can switch on the hardware version.

Reply:
Regarding pushing the version identifiers to the codecs - do you think it will still be helpful when using the `vendorID/vendorProfileID/TBA` schematic ? because then it really becomes blurry what hardware is actually running the formatter / normalizer.